multicluster/controllers/multicluster/leader/member_cleanup_controller.go

```diff
@@ -0,0 +1,199 @@
+	clusterID := getClusterIDFromName(req.Name)
+	staleResExports, err := getResourceExportsByClusterIDFunc(c, ctx, clusterID)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get ResourceExports in the Namespace by ClusterID", "namespace", c.namespace, "clusterID", clusterID)
+		return ctrl.Result{}, err
+	}
+	cleanupSucceed := c.cleanUpResourceExports(ctx, staleResExports)
+	if !cleanupSucceed {
+		return ctrl.Result{}, fmt.Errorf("failed to clean up all stale ResourceExports for the member cluster %s, retry later", clusterID)
+	}
+	return ctrl.Result{}, nil
+}
+
+func getResourceExportsByClusterID(c *MemberResourcesCleanupController, ctx context.Context, clusterID string) ([]mcv1alpha1.ResourceExport, error) {
+	resourceExports := &mcv1alpha1.ResourceExportList{}
+	err := c.Client.List(ctx, resourceExports, &client.ListOptions{Namespace: c.namespace}, client.MatchingFields{indexKey: clusterID})
+	if err != nil {
+		klog.ErrorS(err, "Failed to get ResourceExports by ClusterID", "clusterID", clusterID)
+		return nil, err
+	}
+	return resourceExports.Items, nil
+}
+
+func (c *MemberResourcesCleanupController) cleanUpStaleResourceExports(ctx context.Context) error {
+	existingMemberClusterIDs, err := c.getExistingMemberClusterIDs(ctx)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get existing member cluster's ClusterID in the Namespace", "namespace", c.namespace)
+		return err
+	}
+
+	resourceExports := &mcv1alpha1.ResourceExportList{}
+	err = c.Client.List(ctx, resourceExports, &client.ListOptions{Namespace: c.namespace})
+	if err != nil {
+		klog.ErrorS(err, "Failed to get ResourceExports in the Namespace", "namespace", c.namespace)
+		return err
+	}
+
+	staleResExports := []mcv1alpha1.ResourceExport{}
+	for _, resExport := range resourceExports.Items {
+		// The AntreaClusterNetworkPolicy kind of ResourceExport is created in the leader directly
+		// without a ClusterID info. It's not owned by any member cluster.
+		if resExport.Spec.ClusterID != "" && !existingMemberClusterIDs.Has(resExport.Spec.ClusterID) {
+			staleResExports = append(staleResExports, resExport)
+		}
+	}
+
+	cleanUpSucceed := c.cleanUpResourceExports(ctx, staleResExports)
+	if !cleanUpSucceed {
+		return fmt.Errorf("stale ResourceExports are not fully cleaned up, retry later")
+	}
+	return nil
+}
+
+func (c *MemberResourcesCleanupController) getExistingMemberClusterIDs(ctx context.Context) (sets.Set[string], error) {
+	validMemberClusterIDs := sets.Set[string]{}
+	memberClusterAnnounces := &mcv1alpha1.MemberClusterAnnounceList{}
+	err := c.Client.List(ctx, memberClusterAnnounces, &client.ListOptions{Namespace: c.namespace})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, m := range memberClusterAnnounces.Items {
+		validMemberClusterIDs.Insert(m.ClusterID)
+	}
+	return validMemberClusterIDs, nil
+}
+
+func (c *MemberResourcesCleanupController) cleanUpResourceExports(ctx context.Context, resouceExports []mcv1alpha1.ResourceExport) bool {
```

Comment:
Call it `deleteResourceExports` as the func itself does not really know about cleanup or stale resources?